readme.md
![](banner_big.png)
**Drawfire** is simple drawing application made on [Nimfire framework](https://github.com/Toma400/Nimfire)
as example to learn code from. However, process of writing it was also fun enough
for me to decide to make it an app on its own.  
Don't expect anything groundbreaking - it is just fun side-project and it won't get
anywhere close to legends like Krita or GIMP. It's still fun to play with.

### What features does it have?
It allows you to draw with few predetermined colours, as well as adjust brush size
and type. It exports images to PNG, of strict size 800x600.  
From advantages it has I would definitely say that it is extremely small - so if you
want to draw something really quickly, it will definitely be possible.

Additionally, you can feel safe with Nimfire drawing - it follows "no images lost"
rule which means that you will never mistakingly overwrite any image you work on.  
Whatever you save, is saved as new file, and whatever image you load and save after
editing, is saved also as new file. While this may feel chaotic, it ensures you will
always have as many history points of your image as you want.

![](drawfire_example.png)

### How to use it?
Whenever you run the app, you get empty canvas. Draw with your mouse.  
Here are shortcuts that you may need:

|    Key(s)    | Usage                            |
|:------------:|:---------------------------------|
|    MOUSE     | Use any of mouse buttons to draw |
|     TAB      | Saves the image                  |
|      Q       | Clears the canvas                |
|      T       | Changes transparency of canvas   |
|  UP / DOWN   | Changes size of brush            |
| LEFT / RIGHT | Changes brush type               |
|  `[` / `]`   | Change angle of dynamic brush    |

If you are not sure about something, please refer to [PDF Manual](docs/Drawfire%20Manual.pdf).
It also comes with Drawfire bundle on [releases page](https://github.com/Toma400/Drawfire/releases).

### Feature chart
None of features listed as `todo` here is promised to be completed.
[See changelog for brief overview of changes](changelog.md).

- [x] Basics
  - [x] Changing brush size 
  - [ ] Filling tool
  - [ ] Eraser
- [x] Colour
  - [x] Predefined colour picker 
  - [ ] Fluid colour picker 
- [x] Brushes
  - [x] Notifier of current brush type & size 
  - [x] Various brushes
    - [x] Circle brush 
  - [ ] Custom brushes
    - [ ] JSON-driven
    - [ ] PNG-driven <!-- reading matrix and creating specific brush through it? -->
- [X] Saving the image
- [x] Loading the image
- [ ] Advanced
  - [ ] Layers 
  - [x] Transparency support
- [ ] APIs
  - [ ] Wacom 

<!--
### 0.2.2
[ ] Switchable pallettes? (by default `colours.json` is used, but you can have
    multiple JSONs with different numbers (`colours2.json`, `colours3.json`) and
    you can swap them mid-drawing using new buttons which would reload them?
[ ] Copy to clipboard
[ ] Eraser (?)
    - would just set pencil to background colour and reversely, remembering last
      colour used
[ ] Fix URL linking too often/mistakingly
### 0.2.3
[ ] Proper separating of transparency and background (commented out idea?)
[ ] Fill bucket
### 0.2.4
[ ] JSON-driven or PNG-driven custom brushes?
[ ] "Autosave" mode that saves images each 10/20/30 seconds in separate folder
    and keeps last 50 versions?

### 0.3.0
-- versions before 0.3.0 gets called "Legacy" and are updated only with
   major features in 0.2.n, code gets copy-split into different branch?
[ ] Revamping Drawfire GUI with Dylan's textures? :woa:
[ ] Layers
[ ] HSV picker
[ ] Transparency picker

### 0.4.0
[ ] Wacom support

### 0.5.0
[ ] Ctrl+Z? (would need storing last draws in some sequence of specific size
             but this has capacity/lag issues and also would be hard to control
             without some good precise timer)

### 0.6.0? (very experimental idea)
[ ] Multidimensional/sized images? (bigger/smaller than 800x600?)
[ ] Connected to earlier - glass tool that makes canvas bigger/smaller?

changelog.md
# Versions
- [0.2.2](#0.2.2)
- [0.2.1](#0.2.1)
- [0.2.0](#0.2.0)
- [0.1.0](#0.1.0)

### 0.2.2


### 0.2.1
- Added a way to overwrite default palette by using `colours.json` file
- Added loading of images
- Realigned GUI, so you have full canvas on the left and buttons (with resized
  Drawfire, so it won't distract you with drawing) on the right
- Drawfire logo now redirects you to [Drawfire releases page](https://github.com/Toma400/Drawfire/releases)
  where you can see if there's new update available
- Making whole Drawfire system mouse button-agnostic, meaning full mouse support
  for both left- and right-handed

### 0.2.0
- Created fully GUI-driven app, adding alternative to all key-driven features
  - Brush can now be fully seen in respective window 
- Added transparency switch, allowing you to draw without background
- Increased amount of colours
- Increased amount of brushes
  - Added circle brush
  - Added dynamic brushes, including custom-angled one
- Added use of right-clicking mouse as alternative for drawing
  (can be handy for left-handed people!)
- Brush size was changed to 15 (previously: 1), so it doesn't need scaling up instantly
- Made brush switch slower, so it won't skip brushes

### 0.1.0
- Added basic colours
- Added basic brushes (slash, cross, square)
- Added saving of images

brushes.nim
import std/sequtils
import std/math
# Brush kinds to add
# - algorithm (JSON) driven
# - image (PNG) driven

type
  BRUSHES* = enum
    SLASH, SQUARE, CROSS, CIRCLE, ROTABLE_MANUALLY, ROTABLE_X, ROTABLE_Y, ROTABLE_CROSS

iterator getBrushes: BRUSHES =
    for br in BRUSHES.low..BRUSHES.high:
      yield br
let brushList*  = toSeq(getBrushes())
let brushCount* = brushList.len

proc cycleBrushes* (b: BRUSHES, order: int): BRUSHES =
    proc getIndex (b: BRUSHES): int =
      for ix, brushS in brushList:
        if brushS == b: return ix

    proc setIndex (ix: int, o: int): int =
      if ix+o > brushList.len-1:
        return setIndex(ix, o-brushList.len)
      elif ix+o < 0:
        return setIndex(ix, o+brushList.len)
      else:
        return (ix+o)

    let ix = getIndex(b)
    return brushList[setIndex(ix, order)] # -1 because we aim for index

iterator brushSlash* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      let x = pos[0] + i
      let y = pos[1] + i
      yield (x, y)

iterator brushSquare* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      for j in -brush..brush:
        let x = pos[0] + i
        let y = pos[1] + j
        yield (x, y)

iterator brushCross* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      let x = pos[0] + i
      let y = pos[1]
      yield (x, y)
    for j in -brush..brush:
      let x = pos[0]
      let y = pos[1] + j
      yield (x, y)

iterator brushCircle* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      for j in -brush..brush:
        if i*i + j*j <= brush*brush:
          let x = pos[0] + i
          let y = pos[1] + j
          yield (x, y)

iterator brushRotableManually* (brush: int, pos: (int, int), angle: int): (int, int) =
    for i in -brush..brush:
      let sx = cos(angle.float * PI / 180)
      let sy = sin(angle.float * PI / 180)
      yield ((pos[0].float + sx * i.float).int, (pos[1].float + sy * i.float).int)

iterator brushRotableX* (brush: int, pos: (int, int)): (int, int) =
    let angle = pos[0]
    for i in -brush..brush:
      let sx = cos(angle.float * PI / 180)
      let sy = sin(angle.float * PI / 180)
      yield ((pos[0].float + sx * i.float).int, (pos[1].float + sy * i.float).int)

iterator brushRotableY* (brush: int, pos: (int, int)): (int, int) =
    let angle = pos[1]
    for i in -brush..brush:
      let sx = cos(angle.float * PI / 180)
      let sy = sin(angle.float * PI / 180)
      yield ((pos[0].float + sx * i.float).int, (pos[1].float + sy * i.float).int)

iterator brushRotableCross* (brush: int, pos: (int, int)): (int, int) =
    let angle = pos[0]*pos[1]
    for i in -brush..brush:
      let sx = cos(angle.float * PI / 180)
      let sy = sin(angle.float * PI / 180)
      yield ((pos[0].float + sx * i.float).int, (pos[1].float + sy * i.float).int)

proc brushPick* (kind: BRUSHES, brush: int, pos: (int, int), angle: int): seq[(int, int)] =
    case kind:
      of SLASH:
        result = brushSlash(brush, pos).toSeq
      of SQUARE:
        result = brushSquare(brush, pos).toSeq
      of CROSS:
        result = brushCross(brush, pos).toSeq
      of CIRCLE:
        result = brushCircle(brush, pos).toSeq
      of ROTABLE_X:
        result = brushRotableX(brush, pos).toSeq
      of ROTABLE_Y:
        result = brushRotableY(brush, pos).toSeq
      of ROTABLE_CROSS:
        result = brushRotableCross(brush, pos).toSeq
      of ROTABLE_MANUALLY:
        result = brushRotableManually(brush, pos, angle).toSeq

#[ iterators made by Lyofi for not exactly brushes XD
iterator brushTriangle* (brush: int, pos: (int, int)): (int, int) =  # >
    for i in -brush..brush:
      let x = pos[0]-abs(i)
      let y = pos[1]+i
      yield (x, y)

iterator brushTriangleB* (brush: int, pos: (int, int)): (int, int) =  # >> (stronger >)
    for i in -brush..brush:
      for j in 0..brush:
        let x = pos[0]-abs(i)+j-brush
        let y = pos[1]+i
        yield (x, y)

iterator brushTriangleC* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      for j in 0..abs(i):
        let x = pos[0]+abs(i)+j
        let y = pos[1]+i
        yield (x, y)

iterator brushTriangleD* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      for j in 0..(brush-abs(i)):
        let x = pos[0]+abs(i)+j
        let y = pos[1]+i
        yield (x, y)

iterator brushTriangleE* (brush: int, pos: (int, int)): (int, int) =
    for i in -brush..brush:
      for j in 0..(brush-abs(i)):
        let x = pos[0]-abs(i)+j
        let y = pos[1]+i
        yield (x, y)
iterator brushUnknown* (brush: int, pos: (int, int), angle: int = pos[0]): (int, int) =
    for i in -brush..brush:
      let sx = cos(angle.float * PI / 180)
      let sy = sin(angle.float * PI / 180)
      yield (pos[0] + sx.int * i, pos[1] + sy.int * i)
 ]#

buttons.nim
from chroma import ColorRGBX
import std/private/oscommon
import std/tables
import nimfire/colors
import nimfire/draw
import nimfire/types
import nimfire
import json

var jsCols = newTable[string, seq[int]]()
var dfCols = {  "1": BLUE,
                "2": GREEN,
                "3": RED,
                "4": YELLOW,
                "5": uPASTEL_LIGHT_BLUE,
                "6": LIME,
                "7": ORANGE_RED,
                "8": ACAJOU,
                "9": PURPLE,
               "10": BLACK,
               "11": WHITE,
               "12": GRAY,
               "13": TEAL,
               "14": CREAM,
               "15": BAIKO,
               "16": uGRAYED_CREAM}.toTable

if fileExists("colours.json"):
  try:
    var f = readFile("colours.json")
    var js = parseJson(f)
    for ix in 1..16:
        var jns = js[$ix].getElems()
        var ret = newSeq[int](3)
        for ji in jns.items:
            ret.add(ji.getInt())
        jsCols[$ix] = ret
  except: discard

proc getColors* (s: string): ColorRGBX =
    try: result = toRGBX(jsCols[s][3].uint8,
                         jsCols[s][4].uint8,
                         jsCols[s][5].uint8, 255.uint8)
    except: result = dfCols[s]

proc isTransparent* (c: Rect): bool =
    return c.colour == toRGBX(0, 0, 0, 0)

var blueButton*   = newRect((1020, 20), (50, 50), getColors("1"))
var greenButton*  = newRect((1080, 20), (50, 50), getColors("2"))
var redButton*    = newRect((1140, 20), (50, 50), getColors("3"))
var yellowButton* = newRect((1200, 20), (50, 50), getColors("4"))

var lightBlueButton* = newRect((1020, 80), (50, 50), getColors("5"))
var limeButton*      = newRect((1080, 80), (50, 50), getColors("6"))
var orangeButton*    = newRect((1140, 80), (50, 50), getColors("7"))
var brownButton*     = newRect((1200, 80), (50, 50), getColors("8"))

var purpleButton* = newRect((1020, 140), (50, 50), getColors("9"))
var blackButton*  = newRect((1080, 140), (50, 50), getColors("10"))
var whiteButton*  = newRect((1140, 140), (50, 50), getColors("11"))
var grayButton*   = newRect((1200, 140), (50, 50), getColors("12"))

var cyanButton*  = newRect((1020, 200), (50, 50), getColors("13"))
var creamButton* = newRect((1080, 200), (50, 50), getColors("14"))
var baikoButton* = newRect((1140, 200), (50, 50), getColors("15"))
var crayButton*  = newRect((1200, 200), (50, 50), getColors("16"))

let leftBrushButton* = newRect((820, 220), (50, 50), ACAJOU)
let rightBrushButton* = newRect((950, 220), (50, 50), ACAJOU)

let saveButton*    = newRect((820, 280), (50, 50), ACAJOU)
let saveButtonAE1* = newRect((830, 290), (30, 30), BLUE)
let saveButtonAE2* = newRect((836, 290), (20, 10), GRAY)
let saveButtonAE3* = newRect((842, 291), ( 8,  6), BLUE)

let trButton*    = newRect((880, 280), (50, 50), ACAJOU) # transparency handling
let trButtonAE1* = newRect((891, 291), (20, 20), WHITE)
let trButtonAE2* = newRect((901, 301), (20, 20), BLACK)

let cleanButton*    = newRect((950, 280), (50, 50), ACAJOU)

let separator*        = newRect((800, 382), (480, 5), ACAJOU)
let leftImageButton*  = newRect((965, 410), (50, 50), ACAJOU)
let rightImageButton* = newRect((1065, 410), (50, 50), ACAJOU)
let loadImageButton*  = newRect((1025, 400), (30, 70), ACAJOU)

proc leftBrushButtonArrow* (w: var Window, size: int) =
    let pos = (835, 245)               # brush buttons
    for i in -size..size:
      for j in 0..(size-abs(i)):
        let x = pos[0]+abs(i)+j
        let y = pos[1]+i
        w.fillPos((x, y), uDARK_BROWN)

    let pis = (985, 435)              # image switch buttons
    for i in -size..size:
      for j in 0..(size-abs(i)):
        let x = pis[0]+abs(i)+j
        let y = pis[1]+i
        w.fillPos((x, y), MORNING_BLUE)
proc rightBrushButtonArrow* (w: var Window, size: int) =
    let pos = (985, 245)               # brush buttons
    for i in -size..size:
      for j in 0..(size-abs(i)):
        let x = pos[0]-abs(i)-j
        let y = pos[1]+i
        w.fillPos((x, y), uDARK_BROWN)

    let pis = (1095, 435)              # image switch buttons
    for i in -size..size:
      for j in 0..(size-abs(i)):
        let x = pis[0]-abs(i)-j
        let y = pis[1]+i
        w.fillPos((x, y), MORNING_BLUE)
proc downBrushButtonArrow* (w: var Window, size: int) =
    let pis = (1040, 420)
    for i in -size..size:
      for j in 0..(size-abs(i)):
        let x = pis[0]+i
        let y = pis[1]-abs(i)+j
        w.fillPos((x, y), uGRAYED_CREAM)

    let pvs = (1040, 450)
    for i in -size..size:
      for j in 0..(size-abs(i)):
        let x = pvs[0]+i
        let y = pvs[1]-abs(i)+j
        w.fillPos((x, y), uGRAYED_CREAM)
proc cleanButtonCross* (wn: var Window) =
    proc draw(p: (int, int), s: int, w: var Window) =
      for i in -s..s:     # s = size
        let x = p[0] + i  # p = pos
        let y = p[1] + i
        w.fillPos((x, y), RED)
        let a = p[0] + i
        let b = p[1] - i
        w.fillPos((a, b), RED)
    draw((976, 304), 12, wn)
    draw((974, 304), 12, wn)
    draw((976, 306), 12, wn)
    draw((974, 306), 12, wn)

# lists that should be updated to update drawing boards
let colourButtons* = @[blueButton,
                       greenButton,
                       redButton,
                       yellowButton,
                       lightBlueButton,
                       limeButton,
                       orangeButton,
                       brownButton,
                       purpleButton,
                       blackButton,
                       whiteButton,
                       grayButton,
                       cyanButton,
                       creamButton,
                       baikoButton,
                       crayButton]

drawfire.nim
from pixie/fileformats/png import encodePng
import clipboard
import std/strformat
import std/browsers
import std/sequtils
import std/tables
import std/os
import nimfire/colors
import nimfire/input
import nimfire/image # required for saving canvas as PNG
import nimfire/types # required for 'brushDraw' proc
import nimfire/draw
import nimfire
import brushes
import buttons
import os

proc getBanner(): Image =
  if fileExists("banner.png"): return newImage("banner.png", (800, 480))
  else:                        return newRect((0, 600), (1280, 320), CHOCOLATE).toImage

var w    = initWindow((1280, 600), "Drawfire", bg_colour=CHOCOLATE) # 920
var logo = getBanner()

# var canvbg = newRect(pos=(0, 0), size=(800, 600), WHITE) # background of canvas (if transparency is set, it allows for white background)
var canvas = newRect(pos=(0, 0), size=(800, 600), WHITE) # canvas (for drawing)
var images = newSeq[string]() # will handle

# loaded image
var load = 0
var timer = 0
var angle = 30
var brush = 15 # size of a brush
var brush_type = BRUSHES.SLASH
let brush_rect = newRect(pos=(820, 20), size=(180, 180), IWAI) # background
var brush_kind = newRect(pos=(910, 110), size=(50, 50), BLACK)  # marker showcasing type of brush & colour
var url_timer  = 0 # timer for avoiding re-clicking logo URL

proc brushDraw(w: var Window, pos: (int, int), b: BRUSHES) =
    var coords = brushPick(b, brush, pos, angle)
    for coord in coords:
      w.fillPos(coord, brush_kind.colour)
      setPixelAbsolute(canvas, coord, brush_kind.colour)

proc dynamicBrush(w: var Window) =
    let p = (910, 245)
    let s = 15           # size of icon
    var c = CHOCOLATE    # default colour: background
    if brush_type in [ROTABLE_CROSS, ROTABLE_X, ROTABLE_Y, ROTABLE_MANUALLY]:
      if   brush_type == ROTABLE_CROSS:    c = RED
      elif brush_type == ROTABLE_X:        c = GREEN
      elif brush_type == ROTABLE_Y:        c = BLUE
      elif brush_type == ROTABLE_MANUALLY: c = CYAN
    for x in -s..s:
      if x > 0:
        w.fillPos((p[0]+x, p[1]+x), c)
        w.fillPos((p[0]+x, p[1]-x), c)
      if x == 0:
        for y in -s..s:
          w.fillPos((p[0], p[1]+y), c)
          if y == 0:
            for x2 in -s..s:
              w.fillPos((p[0]+x2, p[1]), c)
      if x < 0:
        w.fillPos((p[0]+x, p[1]+x), c)
        w.fillPos((p[0]+x, p[1]-x), c)

proc cycleImages(mode: int = 0) =
    images = toSeq(walkFiles("image*.png"))
    if mode == 1:
      if images.len > load+1 and load+1 > 0:
        load += 1
        timer = 20
      elif load+1 == images.len:
        load  = 0
        timer = 20
    elif mode == -1:
      # 4 > [-1] && 0 > 0
      if load > 0:
        load -= 1
        timer = 20
      elif load == 0:
        load = images.len-1
        timer = 20
    try: # handling wrong size
      let limg = newImage(images[load])
      if limg.res[0] <= canvas.size[0] and limg.res[1] <= canvas.size[1]:
        canvas.matrix = limg.matrix
        w.drawRect(canvas)
    except: discard

proc save(r: Rect) =
    for i in 1..100_000:
      if not fileExists(fmt"image{i}.png"):
        saveImage(r.toImage(), fmt"image{i}.png")
        break

proc trRedraw(cond: bool = true) =
    if cond:
      if isTransparent(canvas):
        setColour(canvas, WHITE)
      else:
        setColour(canvas, toRGBX(0, 0, 0, 0))
    w.drawRect(trButton)
    if isTransparent(canvas):
      w.drawRect(trButtonAE1)
      w.drawRect(trButtonAE2)
    else:
      w.drawRect(trButtonAE2)
      w.drawRect(trButtonAE1)

proc clear() =
    w.clear()
    clearPixels(canvas)
    # w.drawRect(canvbg)
    w.drawRect(canvas)
    w.drawRect(brush_rect)
    block GUI:
      w.drawRect(leftBrushButton)
      w.drawRect(rightBrushButton)
      w.drawRect(saveButton)
      w.drawRect(saveButtonAE1)
      w.drawRect(saveButtonAE2)
      w.drawRect(saveButtonAE3)
      w.drawRect(cleanButton)
      w.drawRect(separator)
      w.drawRect(leftImageButton)
      w.drawRect(rightImageButton)
      w.drawRect(loadImageButton)
      w.cleanButtonCross()
      w.dynamicBrush()
      trRedraw(false)
      w.leftBrushButtonArrow(10)
      w.rightBrushButtonArrow(10)
      w.downBrushButtonArrow(8)
    block Buttons:
      for cbut in colourButtons:
        w.drawRect(cbut)
    block Images:
      w.drawImage(logo)

clear()

while w.tick():
  brushDraw(w, brush_kind.pos, brush_type)

  if timer     > 0: timer     -= 1 # ticks down timers
  if url_timer > 0: url_timer -= 1

  if w.getKeyPressed(KEY.UP):
    w.drawRect(brush_rect)
    brush += 1
  elif w.getKeyPressed(KEY.DOWN):
    w.drawRect(brush_rect)
    if brush > 1:
      brush -= 1
  elif w.getKeyPressed(KEY.LEFT) and timer == 0:
    w.drawRect(brush_rect)
    brush_type = cycleBrushes(brush_type, -1)
    w.dynamicBrush()
    timer = 20
  elif w.getKeyPressed(KEY.RIGHT) and timer == 0:
    w.drawRect(brush_rect)
    brush_type = cycleBrushes(brush_type, +1)
    w.dynamicBrush()
    timer = 20
  elif w.getKeyPressed(KEY.L_BRACKET):
    w.drawRect(brush_rect)
    angle += -1
  elif w.getKeyPressed(KEY.R_BRACKET):
    w.drawRect(brush_rect)
    angle += 1
  elif w.getKeyPressed(KEY.TAB):
    save(canvas)
  elif w.getKeyPressed(KEY.T):
    trRedraw()
    clear()
  elif w.getKeyPressed(KEY.L_CTRL) or w.getKeyPressed(KEY.R_CTRL):
    discard
  #     let p = clipboardWithName(CboardGeneral) # _H:0001 | _H:0007 | _H:0010 | _H:000D | HTML Format
  #     writeData(p, "_H:0001", encodePng((canvas.toImage).png))
  #     # setClipboardString(w.scr.win.ct, "Hello world".cstring) # encodePng((canvas.toImage).png)
  #     # https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setclipboarddata
  #     # https://learn.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats
  # https://github.com/search?q=clipboard+language%3ANim+&type=code
  # https://github.com/search?q=repo%3Ayglukhov%2Fclipboard%20writeData&type=code
  # https://duckduckgo.com/?q=image+to+byte&ia=web
  # https://duckduckgo.com/?q=base64+into+bytes+nim&ia=answer

  # drawing with mouse
  if w.getMousePressed(LEFT) or w.getMousePressed(RIGHT):
    var pos = w.getMousePos()
    if collide(canvas, pos): # and we also add condition for drawing to do drawing only when it is within canvas
      brushDraw(w, pos, brush_type)
      url_timer = 100
    else:
      if collide(blueButton, pos):        brush_kind.setColour(getColors("1"))
      elif collide(greenButton, pos):     brush_kind.setColour(getColors("2"))
      elif collide(redButton, pos):       brush_kind.setColour(getColors("3"))
      elif collide(yellowButton, pos):    brush_kind.setColour(getColors("4"))
      elif collide(lightBlueButton, pos): brush_kind.setColour(getColors("5"))
      elif collide(limeButton, pos):      brush_kind.setColour(getColors("6"))
      elif collide(orangeButton, pos):    brush_kind.setColour(getColors("7"))
      elif collide(brownButton, pos):     brush_kind.setColour(getColors("8"))
      elif collide(purpleButton, pos):    brush_kind.setColour(getColors("9"))
      elif collide(blackButton, pos):     brush_kind.setColour(getColors("10"))
      elif collide(whiteButton, pos):     brush_kind.setColour(getColors("11"))
      elif collide(grayButton, pos):      brush_kind.setColour(getColors("12"))
      elif collide(cyanButton, pos):      brush_kind.setColour(getColors("13"))
      elif collide(creamButton, pos):     brush_kind.setColour(getColors("14"))
      elif collide(baikoButton, pos):     brush_kind.setColour(getColors("15"))
      elif collide(crayButton, pos):      brush_kind.setColour(getColors("16"))
      else:
        if collide(leftBrushButton, pos) and timer == 0:
          w.drawRect(brush_rect)
          brush_type = cycleBrushes(brush_type, -1)
          w.dynamicBrush()
          timer = 20
        elif collide(rightBrushButton, pos) and timer == 0:
          w.drawRect(brush_rect)
          brush_type = cycleBrushes(brush_type, +1)
          w.dynamicBrush()
          timer = 20
        elif collide(saveButton, pos):
          save(canvas)
        elif collide(leftImageButton, pos) and timer == 0:
          cycleImages(-1)
        elif collide(rightImageButton, pos) and timer == 0:
          cycleImages(1)
        elif collide(loadImageButton, pos) and timer == 0:
          cycleImages()
        elif collide(trButton, pos):
          trRedraw()
          clear()
        elif collide(cleanButton, pos):
          clear()
        elif collide(logo, pos) and url_timer == 0:
          openDefaultBrowser("https://github.com/Toma400/Drawfire/releases")
          url_timer = 50

  w.update(manual=true)

w.finish()